usr/sbin/pkcsstats/pkcsstats.c

```diff
@@ -0,0 +1,867 @@
+    sd.summary_data = calloc(sd.summary_size, 1);
+    if (sd.summary_data == NULL) {
+        warnx("Failed to allocate the summary buffer");
+        return 1;
+    }
+
+    rc = for_all_users(display_summary_cb, &sd);
+    if (rc != 0)
+        goto done;
+
+    printf("Summary (all users):\n\n");
+
+    rc = for_all_slots(display_slot_cb, dd, sd.summary_data, sd.summary_size,
+                       dd->num_slots, dd->slots,
+                       dd->slot_id_specified, dd->slot_id);
+
+done:
+    free(sd.summary_data);
+
+    return rc;
+}
+
+int init_ock(void **dll, CK_FUNCTION_LIST_PTR *func_list)
+{
+    void (*sym_ptr)();
+    CK_RV rc;
+
+    *dll = dlopen("libopencryptoki.so", RTLD_NOW);
+    if (*dll == NULL) {
+        warnx("Error loading PKCS#11 library: dlopen: %s", dlerror());
+        return 1;
+    }
+
+    *(void **)(&sym_ptr) = dlsym(*dll, "C_GetFunctionList");
+    if (sym_ptr == NULL) {
+        warnx("Error loading PKCS#11 library: dlsym(C_GetFunctionList): %s",
+              dlerror());
+        dlclose(*dll);
+        *dll = NULL;
+        return 1;
+    }
+
+    sym_ptr(func_list);
+    if (*func_list == NULL) {
+        warnx("Error getting function list from PKCS11 library");
+        dlclose(*dll);
+        *dll = NULL;
+        return 1;
+    }
+
+    rc = (*func_list)->C_Initialize(NULL);
+    if (rc != CKR_OK) {
+        warnx("Error initializing the PKCS11 library: 0x%lX (%s)", rc,
+               p11_get_ckr(rc));
+        dlclose(*dll);
+        *dll = NULL;
+        *func_list = NULL;
+        return 1;
+    }
+
+    return 0;
+}
+
+int main(int argc, char **argv)
+{
+    int opt = 0;
+    struct passwd *pswd = NULL;
+    int user_id = -1;
+    bool summary = false, all_users = false, all_mechs = false;
+    bool reset = false, reset_all = false;
+    bool delete = false, delete_all = false;
+    bool slot_id_specified = false;
+    CK_SLOT_ID slot_id = 0;
+    void *dll = NULL;
+    CK_FUNCTION_LIST *func_list = NULL;
+    CK_ULONG num_slots = 0;
+    CK_SLOT_ID *slots = NULL;
+    char *endptr;
+    int rc = 0;
+    struct display_data dd;
+    struct reset_data rd;
+
+    static const struct option long_opts[] = {
+        {"user", required_argument, NULL, 'U'},
+        {"summary", no_argument, NULL, 'S'},
+        {"all", no_argument, NULL, 'A'},
+        {"all-mechs", no_argument, NULL, 'a'},
+        {"slot", required_argument, NULL, 's'},
+        {"reset", no_argument, NULL, 'r'},
+        {"reset-all", no_argument, NULL, 'R'},
+        {"delete", no_argument, NULL, 'd'},
+        {"delete-all", no_argument, NULL, 'D'},
+        {"help", no_argument, NULL, 'h'},
+        {0, 0, 0, 0}
+    };
+
+    while ((opt = getopt_long(argc, argv, "U:SAas:rRdDvh", long_opts, NULL)) != -1) {
+        switch (opt) {
+        case 'U':
+            if((pswd = getpwnam(optarg)) == NULL){
+                warnx("The username '%s' is not known on this system.", optarg);
```

Comment:
All other messages don't have a dot at the end ....

Reply:
fixed